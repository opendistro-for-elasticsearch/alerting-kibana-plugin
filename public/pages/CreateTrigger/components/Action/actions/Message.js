/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import _ from 'lodash';
import Mustache from 'mustache';
import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
  EuiText,
  EuiButtonEmpty,
  EuiTextArea,
} from '@elastic/eui';

import {
  FormikTextArea,
  FormikFieldText,
  FormikCheckbox,
  FormikFieldNumber,
} from '../../../../../components/FormControls';
import {
  isInvalid,
  isInvalidActionThrottle,
  validateActionThrottle,
  hasError,
  required,
} from '../../../../../utils/validate';
import { URL, MAX_THROTTLE_VALUE, WRONG_THROTTLE_WARNING } from '../../../../../../utils/constants';

const messageHelpText = (index, sendTestMessage) => (
  <EuiFlexGroup justifyContent="spaceBetween" alignItems="center">
    <EuiFlexItem>
      <EuiText size="xs">
        Embed variables in your message using Mustache templates.{' '}
        <a href={URL.MUSTACHE}>Learn more about Mustache</a>
      </EuiText>
    </EuiFlexItem>
    <EuiFlexItem grow={false}>
      <EuiText size="xs">
        <EuiButtonEmpty
          onClick={() => {
            sendTestMessage(index);
          }}
        >
          Send test message
        </EuiButtonEmpty>
      </EuiText>
    </EuiFlexItem>
  </EuiFlexGroup>
);

const Message = ({
  action,
  context,
  index,
  isSubjectDisabled = false,
  sendTestMessage,
  setFlyout,
}) => {
  let preview = '';
  try {
    preview = Mustache.render(action.message_template.source, context);
  } catch (err) {
    preview = err.message;
    console.error('There was an error rendering mustache template', err);
  }
  return (
    <div>
      {!isSubjectDisabled ? (
        <FormikFieldText
          name={`actions.${index}.subject_template.source`}
          formRow
          fieldProps={{ validate: required }}
          rowProps={{
            label: 'Message subject',
            isInvalid,
            error: hasError,
          }}
          inputProps={{ isInvalid }}
        />
      ) : null}
      <FormikTextArea
        name={`actions.${index}.message_template.source`}
        formRow
        fieldProps={{ validate: required }}
        rowProps={{
          label: (
            <div>
              <span>Message</span>
              <EuiButtonEmpty
                size="s"
                onClick={() => {
                  setFlyout({ type: 'message' });
                }}
              >
                Info
              </EuiButtonEmpty>
            </div>
          ),
          helpText: messageHelpText(index, sendTestMessage),
          style: { maxWidth: '100%' },
          isInvalid,
          error: hasError,
        }}
        inputProps={{
          placeholder: 'Can use mustache templates',
          fullWidth: true,
          isInvalid,
        }}
      />

      <EuiFormRow label="Message preview" style={{ maxWidth: '100%' }}>
        <EuiTextArea
          placeholder="Preview of mustache template"
          fullWidth
          value={preview}
          readOnly
          className="read-only-text-area"
        />
      </EuiFormRow>

      <EuiSpacer size="s" />

      <EuiFormRow
        label={
          <div>
            <span style={{ color: '#343741' }}>Action throttling</span>
            <EuiButtonEmpty
              size="s"
              onClick={() => {
                setFlyout({ type: 'messageFrequency' });
              }}
            >
              Info
            </EuiButtonEmpty>
          </div>
        }
        style={{ maxWidth: '100%' }}
      >
        <EuiFlexGroup direction="column">
          <EuiFlexItem
            grow={false}
            style={{ marginBottom: _.get(action, `throttle_enabled`) ? '0px' : '12px' }}
          >
            <FormikCheckbox
              name={`actions.${index}.throttle_enabled`}
              inputProps={{ label: 'Enable action throttling' }}
            />
          </EuiFlexItem>
          <EuiFlexGroup
            alignItems="center"
            style={{ margin: '0px', display: _.get(action, `throttle_enabled`) ? '' : 'none' }}
          >
            <EuiFlexItem grow={false} style={{ marginRight: '0px' }}>
              <EuiFormRow label="Throttle actions to only trigger every">
                <FormikFieldNumber
                  name={`actions.${index}.throttle.value`}
                  fieldProps={{ validate: validateActionThrottle(action) }}
                  formRow={true}
                  rowProps={{
                    isInvalid: isInvalidActionThrottle(action),
                    error: [WRONG_THROTTLE_WARNING],
                  }}
                  inputProps={{
                    style: { width: '400px', height: '40px' },
                    min: 1,
                    max: MAX_THROTTLE_VALUE,
                    compressed: true,
                    append: (
                      <EuiText
                        style={{
                          height: '40px',
                          lineHeight: '24px',
                          backgroundColor: 'transparent',
                          paddingLeft: '2px',
                        }}
                      >
                        minutes
                      </EuiText>
                    ),
                    class: 'euiFieldText',
                    disabled: !_.get(action, `throttle_enabled`) ? 'disabled' : '',
                  }}
                />
              </EuiFormRow>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexGroup>
      </EuiFormRow>
    </div>
  );
};

export default Message;
