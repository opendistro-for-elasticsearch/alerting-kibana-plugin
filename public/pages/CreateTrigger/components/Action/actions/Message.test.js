/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { render,shallow,mount } from 'enzyme';
import { Formik } from 'formik'
import Message from './Message';
import Mustache from 'mustache';

jest.mock('@elastic/eui/lib/components/form/form_row/make_id', () => () => 'testing-id');

function getRenderWrapper(customProps = {}) {
  return render(
    <Formik
      render={() => <Message 
        action={{
          "message_template": {
            "source": "Monitor {{ctx.monitor.name}} just entered alert status. Please investigate the issue.\n- Trigger: {{ctx.trigger.name}}\n- Severity: {{ctx.trigger.severity}}\n- Period start: {{ctx.periodStart}}\n- Period end: {{ctx.periodEnd}}",
            "lang": "mustache"
          }
        }}
        context={{}}
        index={0}
        sendTestMessage={jest.fn()}
        setFlyout={jest.fn()}
      />}
    />
  );
}

describe('Message', () => {
  test('renders', () => {
    const wrapper = getRenderWrapper();
    expect(wrapper).toMatchSnapshot();
  });  

});
