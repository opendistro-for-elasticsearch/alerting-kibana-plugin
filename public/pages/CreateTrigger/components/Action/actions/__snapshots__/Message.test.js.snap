// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message renders 1`] = `
<div>
  <div
    class="euiFormRow"
    id="actions.0.subject_template.source-row"
  >
    <div>
      <label
        aria-invalid="false"
        class="euiFormLabel"
        for="actions.0.subject_template.source"
      >
        Message subject
      </label>
    </div>
    <div
      class="euiFormControlLayout"
    >
      <div
        class="euiFormControlLayout__childrenWrapper"
      >
        <input
          class="euiFieldText"
          name="actions.0.subject_template.source"
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="actions.0.message_template.source-row"
    style="max-width:100%"
  >
    <div>
      <label
        aria-invalid="false"
        class="euiFormLabel"
        for="actions.0.message_template.source"
      >
        <div>
          <span>
            Message
          </span>
          <button
            class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--small"
            type="button"
          >
            <span
              class="euiButtonEmpty__content"
            >
              <span
                class="euiButtonEmpty__text"
              >
                Info
              </span>
            </span>
          </button>
        </div>
      </label>
    </div>
    <textarea
      class="euiTextArea euiTextArea--resizeVertical euiTextArea--fullWidth"
      name="actions.0.message_template.source"
      placeholder="Can use mustache templates"
      rows="6"
    />
    <div
      class="euiFormHelpText euiFormRow__text"
      id="actions.0.message_template.source-help"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiText euiText--extraSmall"
          >
            Embed variables in your message using Mustache templates. 
            <a
              href="https://mustache.github.io/mustache.5.html"
            >
              Learn more about Mustache
            </a>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiText euiText--extraSmall"
          >
            <button
              class="euiButtonEmpty euiButtonEmpty--primary"
              type="button"
            >
              <span
                class="euiButtonEmpty__content"
              >
                <span
                  class="euiButtonEmpty__text"
                >
                  Send test message
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFormRow"
    id="testing-id-row"
    style="max-width:100%"
  >
    <div>
      <label
        class="euiFormLabel"
        for="testing-id"
      >
        Message preview
      </label>
    </div>
    <textarea
      class="euiTextArea euiTextArea--resizeVertical euiTextArea--fullWidth read-only-text-area"
      id="testing-id"
      placeholder="Preview of mustache template"
      readonly=""
      rows="6"
    >
      Monitor  just entered alert status. Please investigate the issue.
- Trigger: 
- Severity: 
- Period start: 
- Period end: 
    </textarea>
  </div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiFormRow"
    id="testing-id-row"
    style="max-width:100%"
  >
    <div>
      <label
        class="euiFormLabel"
        for="testing-id"
      >
        <div>
          <span
            style="color:#343741"
          >
            Action throttling
          </span>
          <button
            class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--small"
            type="button"
          >
            <span
              class="euiButtonEmpty__content"
            >
              <span
                class="euiButtonEmpty__text"
              >
                Info
              </span>
            </span>
          </button>
        </div>
      </label>
    </div>
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
      id="testing-id"
    >
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
        style="margin-bottom:12px"
      >
        <div
          class="euiCheckbox"
        >
          <input
            class="euiCheckbox__input"
            id="actions.0.throttle_enabled"
            name="actions.0.throttle_enabled"
            type="checkbox"
          />
          <div
            class="euiCheckbox__square"
          />
          <label
            class="euiCheckbox__label"
            for="actions.0.throttle_enabled"
          >
            Enable action throttling
          </label>
        </div>
      </div>
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="margin:0px;display:none"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
          style="margin-right:0px"
        >
          <div
            class="euiFormRow"
            id="testing-id-row"
          >
            <div>
              <label
                class="euiFormLabel"
                for="testing-id"
              >
                Throttle actions to only trigger every
              </label>
            </div>
            <div
              class="euiFormRow"
              id="actions.0.throttle.value-row"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--compressed euiFormControlLayout--group"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <input
                    class="euiFieldNumber euiFieldNumber--compressed euiFieldNumber--inGroup"
                    disabled=""
                    max="1440"
                    min="1"
                    name="actions.0.throttle.value"
                    style="width:400px;height:40px"
                    type="number"
                  />
                </div>
                <div
                  class="euiText euiText--medium euiFormControlLayout__append"
                  style="height:40px;line-height:24px;background-color:transparent;padding-left:4px"
                >
                  minutes
                </div>
              </div>
              <div
                class="euiFormHelpText euiFormRow__text"
                id="actions.0.throttle.value-help"
              >
                Throttle value must be greater than 0 and less than 1440
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
